services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_v2/_play_context.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Optional
+
+from fastapi import FastAPI
+
+from ._context_base import ContextInterface, ContextIOInterface, ReservedContextKeys
+from ._errors import (
+    GetTypeMismatchError,
+    NotAllowedContextKeyError,
+    NotInContextError,
+    SetTypeMismatchError,
+)
+from ._models import PlayName, SceneName
+
+
+def _ensure_type_matches(key: str, existing_value: Any, value: Any) -> None:
+    # if a value previously existed,
+    # ensure it has the same type
+    existing_type = type(existing_value)
+    value_type = type(value)
+    if existing_type != value_type:
+        raise SetTypeMismatchError(
+            key=key,
+            existing_value=existing_value,
+            existing_type=existing_type,
+            new_value=value,
+            new_type=value_type,
+        )
+
+
+class PlayContext(ContextIOInterface):
+    """
+    Data container responsible for keeping track of the state of a play.
+    """
+
+    def __init__(
+        self,
+        context: ContextInterface,
+        app: FastAPI,
+        play_name: PlayName,
+        scene_name: SceneName,
+    ) -> None:
+        self._context = context
+        self._app = app
+        self._play_name = play_name
+        self._scene_name = scene_name
+
+        self._local_storage: dict[str, Any] = {}
+
+    async def set(self, key: str, value: Any, *, set_reserved: bool = False) -> None:
```

Comment:
You obviously cannot annotate ``ReserveContextKeys`` because is a class. I mean annotating the parameter ``def set(self, key: ReserveContextKeys, ... )``

but if you say it can really be any key... then yes, leave it as str